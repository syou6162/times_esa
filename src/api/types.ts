import type { HttpsCallableResult } from 'firebase/functions';

// リクエスト/レスポンスの型定義
export type DailyReportRequest = {
  category: string;
}

export type DailyReportResponse = {
  updatedAt: string;
  url: string;
  bodyMd: string;
  bodyHtml: string;
  tags: string[];
  name: string;
  category: string;
}

export type TagListRequest = {}

export type TagListResponse = {
  tags: Array<{
    name: string;
    postsCount: number;
  }>;
  totalCount: number;
}

export type SubmitTextToEsaRequest = {
  category: string;
  tags: string[];
  title: string;
  text: string;
}

export type SubmitTextToEsaResponse = {
  updatedAt: string;
  url: string;
  bodyMd: string;
  bodyHtml: string;
  tags: string[];
  name: string;
  category: string;
}

export type RecentDailyReportsRequest = {}

export type RecentDailyReportsResponse = {
  reports: Array<{
    date: string;
    url: string;
    posts_count: number;
  }>;
}

// APIクライアントのインターフェース
export interface ApiClient {
  getDailyReport(category: string): Promise<HttpsCallableResult<DailyReportResponse>>;
  getTagList(): Promise<HttpsCallableResult<TagListResponse>>;
  submitTextToEsa(
    category: string,
    tags: string[],
    title: string,
    text: string
  ): Promise<HttpsCallableResult<SubmitTextToEsaResponse>>;
  getRecentDailyReports(): Promise<HttpsCallableResult<RecentDailyReportsResponse>>;
}