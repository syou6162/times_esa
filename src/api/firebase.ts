import { getFunctions, httpsCallable, HttpsCallableResult } from 'firebase/functions';
import { functionsRegion } from '../util';

// 型定義
export type DailyReportRequest = {
  category: string;
}

export type DailyReportResponse = {
  updated_at: string;
  url: string;
  body_md: string;
  body_html: string;
  tags: string[];
  name: string;
  category: string;
}

export type TagListRequest = {}

export type TagListResponse = {
  tags: Array<{
    name: string;
    posts_count: number;
  }>;
}

export type SubmitTextToEsaRequest = {
  category: string;
  tags: string[];
  title: string;
  text: string;
}

export type SubmitTextToEsaResponse = {
  updated_at: string;
  url: string;
  body_md: string;
  body_html: string;
  tags: string[];
  name: string;
  category: string;
}

export type RecentDailyReportsRequest = {}

export type RecentDailyReportsResponse = {
  reports: Array<{
    date: string;
    url: string;
    posts_count: number;
  }>;
}

// Firebase Functions呼び出しヘルパー
const callFirebaseFunction = <TRequest, TResponse>(
  functionName: string,
  data?: TRequest,
  options?: { timeout?: number }
): Promise<HttpsCallableResult<TResponse>> => {
  if (!import.meta.env.VITE_API_KEY) {
    throw new Error('Firebase configuration is missing. Please set up Firebase credentials.');
  }
  
  const functions = getFunctions();
  functions.region = functionsRegion;
  
  const callable = httpsCallable<TRequest, TResponse>(
    functions,
    functionName,
    options
  );
  
  return callable(data as TRequest);
};

// Firebase Functions API
export const firebaseApi = {
  getDailyReport: (category: string) => 
    callFirebaseFunction<DailyReportRequest, DailyReportResponse>(
      'dailyReport',
      { category }
    ),
  
  getTagList: () => 
    callFirebaseFunction<TagListRequest, TagListResponse>('tagList'),
  
  submitTextToEsa: (category: string, tags: string[], title: string, text: string) =>
    callFirebaseFunction<SubmitTextToEsaRequest, SubmitTextToEsaResponse>(
      'submitTextToEsa',
      { category, tags, title, text },
      { timeout: 10000 }
    ),
  
  getRecentDailyReports: () =>
    callFirebaseFunction<RecentDailyReportsRequest, RecentDailyReportsResponse>(
      'recentDailyReports'
    ),
};