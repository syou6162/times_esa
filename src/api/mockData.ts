import { format } from 'date-fns';

// 日報のモックデータ
export const mockDailyReportData = {
  updatedAt: new Date().toISOString(),
  url: 'https://example.esa.io/posts/12345',
  bodyMd: `<a id="1000" href="#1000">10:00</a> 朝会に参加しました

---

<a id="1030" href="#1030">10:30</a> タスクAの実装を開始

---

<a id="1200" href="#1200">12:00</a> お昼休憩

---

<a id="1300" href="#1300">13:00</a> タスクAの実装完了、PRを作成

---

<a id="1500" href="#1500">15:00</a> コードレビューの対応

---

<a id="1700" href="#1700">17:00</a> 明日のタスクの準備

---
`,
  bodyHtml: `<p><a id="1000" href="#1000">10:00</a> 朝会に参加しました</p>
<hr>
<p><a id="1030" href="#1030">10:30</a> タスクAの実装を開始</p>
<hr>
<p><a id="1200" href="#1200">12:00</a> お昼休憩</p>
<hr>
<p><a id="1300" href="#1300">13:00</a> タスクAの実装完了、PRを作成</p>
<hr>
<p><a id="1500" href="#1500">15:00</a> コードレビューの対応</p>
<hr>
<p><a id="1700" href="#1700">17:00</a> 明日のタスクの準備</p>
<hr>`,
  tags: ['月曜日', '開発', 'タスクA'],
  name: '日報',
  category: `日報/${format(new Date(), 'yyyy/MM/dd')}`,
};

// 日報が存在しない場合のエラー
export const mockDailyReportNotFoundError = {
  code: 'NOT_FOUND',
  message: 'Daily report not found',
};

// タグ一覧のモックデータ
export const mockTagListData = {
  tags: [
    { name: '開発', postsCount: 156 },
    { name: 'ミーティング', postsCount: 89 },
    { name: 'レビュー', postsCount: 124 },
    { name: 'タスクA', postsCount: 23 },
    { name: 'タスクB', postsCount: 18 },
    { name: 'リファクタリング', postsCount: 45 },
    { name: '調査', postsCount: 67 },
    { name: 'ドキュメント', postsCount: 34 },
    { name: '月曜日', postsCount: 52 },
    { name: '火曜日', postsCount: 51 },
    { name: '水曜日', postsCount: 53 },
    { name: '木曜日', postsCount: 50 },
    { name: '金曜日', postsCount: 49 },
  ],
  totalCount: 13
};

// 最近の日報一覧のモックデータ
export const mockRecentDailyReportsData = {
  reports: [
    {
      date: format(new Date(Date.now() - 1 * 24 * 60 * 60 * 1000), 'yyyy-MM-dd'),
      title: '日報',
      tags: [],
      category: `日報/${format(new Date(Date.now() - 1 * 24 * 60 * 60 * 1000), 'yyyy/MM/dd')}`,
      updatedAt: new Date(Date.now() - 1 * 24 * 60 * 60 * 1000).toISOString(),
      number: 12344,
    },
    {
      date: format(new Date(Date.now() - 2 * 24 * 60 * 60 * 1000), 'yyyy-MM-dd'),
      title: '日報',
      tags: ['開発'],
      category: `日報/${format(new Date(Date.now() - 2 * 24 * 60 * 60 * 1000), 'yyyy/MM/dd')}`,
      updatedAt: new Date(Date.now() - 2 * 24 * 60 * 60 * 1000).toISOString(),
      number: 12343,
    },
    {
      date: format(new Date(Date.now() - 3 * 24 * 60 * 60 * 1000), 'yyyy-MM-dd'),
      title: '日報',
      tags: ['ミーティング', 'レビュー'],
      category: `日報/${format(new Date(Date.now() - 3 * 24 * 60 * 60 * 1000), 'yyyy/MM/dd')}`,
      updatedAt: new Date(Date.now() - 3 * 24 * 60 * 60 * 1000).toISOString(),
      number: 12342,
    },
    {
      date: format(new Date(Date.now() - 4 * 24 * 60 * 60 * 1000), 'yyyy-MM-dd'),
      title: '日報',
      tags: ['開発', 'タスクA'],
      category: `日報/${format(new Date(Date.now() - 4 * 24 * 60 * 60 * 1000), 'yyyy/MM/dd')}`,
      updatedAt: new Date(Date.now() - 4 * 24 * 60 * 60 * 1000).toISOString(),
      number: 12341,
    },
    {
      date: format(new Date(Date.now() - 5 * 24 * 60 * 60 * 1000), 'yyyy-MM-dd'),
      title: '日報',
      tags: [],
      category: `日報/${format(new Date(Date.now() - 5 * 24 * 60 * 60 * 1000), 'yyyy/MM/dd')}`,
      updatedAt: new Date(Date.now() - 5 * 24 * 60 * 60 * 1000).toISOString(),
      number: 12340,
    },
  ],
  totalCount: 5
};

// 投稿時のレスポンスモック（既存の日報に追記される想定）
export const createSubmitResponse = (category: string, tags: string[], title: string, text: string) => {
  const currentBody = mockDailyReportData.bodyMd;
  const newBodyMd = currentBody + text;
  const newBodyHtml = newBodyMd.replace(/\n/g, '<br>').replace(/---/g, '<hr>');

  return {
    updatedAt: new Date().toISOString(),
    url: mockDailyReportData.url,
    bodyMd: newBodyMd,
    bodyHtml: newBodyHtml,
    tags: [...new Set([...tags, ...mockDailyReportData.tags])], // 重複を除去
    name: title,
    category: category,
  };
};
