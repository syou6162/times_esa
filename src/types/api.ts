/**
 * API関連の型定義
 * 
 * APIのリクエスト・レスポンスに関する型を定義
 * 基本的にAPIの仕様に合わせてsnake_caseを使用
 */

// 日報取得API
export type DailyReportRequestType = {
  category: string;
};

export type DailyReportResponseType = {
  updated_at: string;
  url: string;
  body_md: string;
  body_html: string;
  tags: string[];
  name: string;
  category: string;
};

// タグ一覧取得API
export type TagListRequestType = Record<string, never>;

export type TagListResponseType = {
  tags: Array<{
    name: string;
    posts_count: number;
  }>;
};

// 最近の日報一覧取得API
export type RecentDailyReportsRequest = {
  days?: number; // 過去何日分を取得するか（デフォルト: 10）
};

export type RecentDailyReportsResponse = {
  reports: Array<{
    date: string;
    title: string;
    posts_count: number;
    tags: string[];
    formatted_date: string;
    url?: string;
  }>;
  total_count: number;
};

// ESA投稿API
export type SubmitTextToEsaRequestType = {
  category: string;
  tags: string[];
  title: string;
  text: string;
};

export type SubmitTextToEsaResponseType = {
  updated_at: string;
  url: string;
  body_md: string;
  body_html: string;
  tags: string[];
  name: string;
  category: string;
};