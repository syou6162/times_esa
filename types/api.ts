// API レスポンスの共通型定義
// バックエンドとフロントエンドで共有される型

// Esa APIレスポンスの基本型（キャメルケース）
export type EsaPost = {
  bodyMd: string;
  bodyHtml: string;
  number: number;
  name: string;
  tags: string[];
  updatedAt: string;
  url: string;
  category: string;
}

export type EsaTag = {
  name: string;
  postsCount: number;
}

export type EsaTags = {
  tags: EsaTag[];
  totalCount: number;
}

// 日報サマリー関連の型定義
export type DailyReportSummary = {
  date: string;  // yyyy-MM-dd形式
  title: string;
  category: string;
  updatedAt: string;
  number: number;
}

export type RecentDailyReportsResponse = {
  reports: DailyReportSummary[];
  totalCount: number;
}

// APIリクエスト/レスポンスの型定義
export type DailyReportRequest = {
  category: string;
}

export type DailyReportResponse = EsaPost;

export type TagListRequest = {}

export type TagListResponse = EsaTags;

export type SubmitTextToEsaRequest = {
  category: string;
  tags: string[];
  title: string;
  text: string;
}

export type SubmitTextToEsaResponse = EsaPost;

export type RecentDailyReportsRequest = {
  days?: number;
}