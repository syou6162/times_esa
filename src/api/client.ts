import { config } from '../config';
import { FirebaseApiClient } from './clients/firebase';
import { MockApiClient } from './clients/mock';
import type { ApiClient } from './types';

// 環境に応じて適切なクライアントを作成
const createApiClient = (): ApiClient => {
  if (config.useMockApi) {
    return new MockApiClient();
  }
  return new FirebaseApiClient();
};

// シングルトンインスタンス
export const apiClient = createApiClient();