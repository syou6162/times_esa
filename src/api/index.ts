import { apiClient } from './client';

// API関数をエクスポート（使用側は変更不要）
export const getDailyReport = (category: string) => {
  return apiClient.getDailyReport(category);
};

export const getTagList = () => {
  return apiClient.getTagList();
};

export const submitTextToEsa = (
  category: string,
  tags: string[],
  title: string,
  text: string
) => {
  return apiClient.submitTextToEsa(category, tags, title, text);
};

export const getRecentDailyReports = () => {
  return apiClient.getRecentDailyReports();
};

// 型のエクスポート（コンポーネントで使用）
export type {
  DailyReportResponse,
  TagListResponse,
  SubmitTextToEsaResponse,
  RecentDailyReportsResponse,
} from './types';
