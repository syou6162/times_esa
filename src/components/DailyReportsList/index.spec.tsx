// @vitest-environment jsdom
import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import { vi, describe, it, expect, beforeEach } from 'vitest';
import { DailyReportsList } from './index';
import { apiClient } from '../../api/client';

// APIクライアントをモック
vi.mock('../../api/client', () => ({
  apiClient: {
    getRecentDailyReports: vi.fn(),
  },
}));

describe('DailyReportsList', () => {
  const mockOnDateSelect = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('ローディング中はスケルトンを表示すること', () => {
    vi.mocked(apiClient.getRecentDailyReports).mockImplementation(
      () => new Promise(() => {}) // 永遠に待つPromise
    );

    render(<DailyReportsList onDateSelect={mockOnDateSelect} />);

    // Skeletonコンポーネントは特定のroleを持たないため、クラス名で確認
    const skeletons = document.querySelectorAll('.MuiSkeleton-root');
    expect(skeletons.length).toBe(5);
  });

  it('日報の一覧を表示すること', async () => {
    const mockReports = [
      {
        date: '2025-06-20',
        title: '開発作業',
        tags: [],
        category: '日報/2025/06/20',
        updatedAt: '2025-06-20T10:00:00Z',
        number: 123,
      },
      {
        date: '2025-06-19',
        title: 'ミーティング',
        tags: [],
        category: '日報/2025/06/19',
        updatedAt: '2025-06-19T10:00:00Z',
        number: 122,
      },
    ];

    vi.mocked(apiClient.getRecentDailyReports).mockResolvedValueOnce({
      data: { reports: mockReports, totalCount: 2 },
    });

    render(<DailyReportsList onDateSelect={mockOnDateSelect} />);

    await waitFor(() => {
      expect(screen.getByText('2025-06-20(金)')).toBeDefined();
      expect(screen.getByText('開発作業')).toBeDefined();
      expect(screen.getByText('2025-06-19(木)')).toBeDefined();
      expect(screen.getByText('ミーティング')).toBeDefined();
    });
  });

  it('エラー時はエラーメッセージを表示すること', async () => {
    vi.mocked(apiClient.getRecentDailyReports).mockRejectedValueOnce(
      new Error('Network error')
    );

    render(<DailyReportsList onDateSelect={mockOnDateSelect} />);

    await waitFor(() => {
      expect(screen.getByText('日報の一覧を取得できませんでした')).toBeDefined();
    });
  });

  it('日報がない場合はメッセージを表示すること', async () => {
    vi.mocked(apiClient.getRecentDailyReports).mockResolvedValueOnce({
      data: { reports: [], totalCount: 0 },
    });

    render(<DailyReportsList onDateSelect={mockOnDateSelect} />);

    await waitFor(() => {
      expect(screen.getByText('過去の日報がありません')).toBeDefined();
    });
  });

  it('日報をクリックするとonDateSelectが呼ばれること', async () => {
    const mockReports = [
      {
        date: '2025-06-20',
        title: '開発作業',
        tags: [],
        category: '日報/2025/06/20',
        updatedAt: '2025-06-20T10:00:00Z',
        number: 123,
      },
    ];

    vi.mocked(apiClient.getRecentDailyReports).mockResolvedValueOnce({
      data: { reports: mockReports, totalCount: 1 },
    });

    render(<DailyReportsList onDateSelect={mockOnDateSelect} />);

    await waitFor(() => {
      expect(screen.getByText('2025-06-20(金)')).toBeDefined();
    });

    fireEvent.click(screen.getByText('2025-06-20(金)'));
    expect(mockOnDateSelect).toHaveBeenCalledWith('2025-06-20', { title: '開発作業', tags: [] });
  });

  it('選択された日付がハイライトされること', async () => {
    const mockReports = [
      {
        date: '2025-06-20',
        title: '開発作業',
        tags: [],
        category: '日報/2025/06/20',
        updatedAt: '2025-06-20T10:00:00Z',
        number: 123,
      },
    ];

    vi.mocked(apiClient.getRecentDailyReports).mockResolvedValueOnce({
      data: { reports: mockReports, totalCount: 1 },
    });

    const { rerender } = render(
      <DailyReportsList selectedDate="2025-06-20" onDateSelect={mockOnDateSelect} />
    );

    await waitFor(() => {
      const button = screen.getByRole('button', { name: /2025-06-20\(金\)/ });
      expect(button.className).toContain('Mui-selected');
    });
  });


});