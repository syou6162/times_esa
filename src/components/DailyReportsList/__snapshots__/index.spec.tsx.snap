// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DailyReportsList > スナップショットテスト 1`] = `
<ul
  class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
>
  <li
    class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
  >
    <div
      class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
        >
          2025-06-20
        </span>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
        >
          開発作業
        </p>
      </div>
    </div>
  </li>
</ul>
`;
